### PR TITLE
Readability

### PR DESCRIPTION
**Pull request by Trenly**
@OfficialEsco  I haven't fully tested this yet, but I'm 90% sure it will work. 

I added a function just to make things more readable. 
```raw
String.IsValid {String} [-MinLength] [-MaxLength] [-MatchPattern]
  Returns True if and only if
    - String.Length >= MinLength  AND
    - String.Length <= MaxLength AND
    - String -match MatchPattern
```

I also did abstract everything to a hashtable just to keep things readable and for making future updates easier

### CODE REVIEW

**Comment by OfficialEsco** [comment]
Hahaha you just do the things don't you, glad you find inspiration in my ideas 😁
Sure it does kinda look better 🙄

From looking at the code you might have broken the $Max variable on line 591, 596, 601 and 606.

FileExtensions also have a Pattern, but i don't think we can implement it as its multi string?

**Comment by Trenly** [comment]
> Hahaha you just do the things don't you, glad you find inspiration in my ideas 😁
> Sure it does kinda look better 🙄
> 
> From looking at the code you might have broken the $Max variable on line 591, 596, 601 and 606.
> 
> FileExtensions also have a Pattern, but i don't think we can implement it as its multi string?
> 
> Edit: Yupp
> 
> ```
> [Optional] Enter any File Extensions the application could support. For example: html, htm, url (Max )
> FileExtensions:
> 
> [Optional] Enter any Protocols the application provides a handler for. For example: http, https (Max )
> Protocols:
> 
> [Optional] Enter any Commands or aliases to run the application. For example: msedge (Max )
> Commands:
> 
> [Optional] List of additional non-zero installer success exit codes other than known default values by winget (Max )
> InstallerSuccessCodes:
> ```

Oop; Should be an easy fix. I’ll do it when I get home.

We could probably do file extensions enum, I think we already do something similar for Architecture

**Comment by OfficialEsco** [comment]
> We could probably do file extensions enum, I think we already do something similar for Architecture

There is no enum for FileExtensions, however we should probably start looking at error messages so it shows why it failed..
```
"FileExtensions": {
      "type": [ "array", "null" ],
      "items": {
        "type": "string",
        "pattern": "^[^\\\\/:\\*\\?\"<>\\|\\x01-\\x1f]+$",
        "maxLength": 64
      },
      "maxItems": 256,
      "uniqueItems": true,
      "description": "List of file extensions the package could support"
    }
```

```
},
        "Architecture": {
          "type": "string",
          "enum": [
            "x86",
            "x64",
            "arm",
            "arm64",
            "neutral"
          ],
          "description": "The installer target architecture"
        }
```

**Comment by Trenly** [comment]
>Edit2: AHHH yeah you need to look into the MaxItems issue because last time i tried it your PromptInstallerManifestValue function restricted me

What do you mean?

**Comment by OfficialEsco** [comment]
> > Edit2: AHHH yeah you need to look into the MaxItems issue because last time i tried it your PromptInstallerManifestValue function restricted me
> 
> What do you mean?

The current issue is that you can't do `Write-Host "Text here $Variable.MaxItems"` you have to do `Write-Host "Text here", $Variable.MaxItems` but it got errors when i did that

**Comment by Trenly** [comment]
> > > Edit2: AHHH yeah you need to look into the MaxItems issue because last time i tried it your PromptInstallerManifestValue function restricted me
> > 
> > 
> > What do you mean?
> 
> The current issue is that you can't do `Write-Host "Text here $Variable.MaxItems"` you have to do `Write-Host "Text here", $Variable.MaxItems` but it got errors when i did that

The `$( )` operator is really helpful

**Comment by OfficialEsco** [comment]
Interesting.. learned something new, now i know the proper use for `@( )` and `$( )` 👀

File: Tools/YamlCreate.ps1 [hunk 1]
```diff
@@ -510,46 +585,42 @@ Function SortYamlKeys {
...
-    } while (($script:Commands -split ", ").Count -gt $MaxItemsCommands)
+        $script:Commands = PromptInstallerManifestValue $Commands 'Commands' "[Optional] Enter any Commands or aliases to run the application. For example: msedge (Max $($Patterns.MaxItemsCommands))"
+    } while (($script:Commands -split ",").Count -gt $Patterns.MaxItemsCommands)
 
     do {
         if (!$InstallerSuccessCodes) { $InstallerSuccessCodes = '' }
-        $script:InstallerSuccessCodes = PromptInstallerManifestValue $InstallerSuccessCodes 'InstallerSuccessCodes' "[Optional] List of additional non-zero installer success exit codes other than known default values by winget (Max $MaxItemsInstallerSuccessCodes)"
-    } while (($script:InstallerSuccessCodes -split ", ").Count -gt $MaxItemsInstallerSuccessCodes)
+        $script:InstallerSuccessCodes = PromptInstallerManifestValue $InstallerSuccessCodes 'InstallerSuccessCodes' "[Optional] List of additional non-zero installer success exit codes other than known default values by winget (Max $($Patterns.MaxItemsInstallerSuccessCodes))"
```
Comment (OfficialEsco) [inline Tools/YamlCreate.ps1:606]:
```
[Optional] List of additional non-zero installer success exit codes other than known default values by winget (Max )
InstallerSuccessCodes:
```

Comment (Trenly) [reply Tools/YamlCreate.ps1]: Forgot I renamed it. Should be fixed now

**Comment by Trenly** [comment]
> > We could probably do file extensions enum, I think we already do something similar for Architecture
> 
> There is no enum for FileExtensions, however we should probably start looking at error messages so it shows why it failed..
> 
> ```
> "FileExtensions": {
>       "type": [ "array", "null" ],
>       "items": {
>         "type": "string",
>         "pattern": "^[^\\\\/:\\*\\?\"<>\\|\\x01-\\x1f]+$",
>         "maxLength": 64
>       },
>       "maxItems": 256,
>       "uniqueItems": true,
>       "description": "List of file extensions the package could support"
>     }
> ```
> 
> Edit:
> We can probably look into enum for InstallModes first, might be a easier path to finding out how to do pattern for FileExtensions
> 
> ```
> "InstallModes": {
>       "type": [ "array", "null" ],
>       "items": {
>         "title": "InstallModes",
>         "type": "string",
>         "enum": [
>           "interactive",
>           "silent",
>           "silentWithProgress"
>         ]
>       },
> ```
> 

I think I took care of this validation. I don't have it removing duplicate extensions though. Maybe that's something to look at later. I agree that error messages should be next

**Comment by OfficialEsco** [comment]
Was going to report a issue with the InstallModes but then i realized it was a typo, so it clearly works 😁
Probably not getting to stresstest before next week as there are way too few packages on the weekend

**Comment by OfficialEsco** [comment]
@quhxl since you go trough more packages than only GitHub releases and add Metadata would you mind testing this?
Not sure if you've updated/tested the YamlCreate since you've been gone for a few days, we appreciate any feedback :)

**Review by OfficialEsco — APPROVED** [review]

**Pull request merged**